### PR TITLE
Proper documentation

### PR DESCRIPTION
**Pull request by rossviljoen**
Sets up docs CI and cleans up documentation/examples in general.

Heavily pilfered from KernelFunctions and AbstractGPs

[Preview](https://juliagaussianprocesses.github.io/SparseGPs.jl/previews/PR32/)

### CODE REVIEW

**Review by st-- — COMMENTED** [review]
NB: I'm not looking at the files that look like you simply copied them over from the other repos - please advise if there's anything you changed in there!

File: examples/regression/script.jl [hunk 1]
```diff
@@ -0,0 +1,196 @@
+# A recreation of <https://gpflow.readthedocs.io/en/master/notebooks/advanced/gps_for_big_data.html>
+
+# # Stochastic Variational Regression
+#
+# In this example, we show how to construct and train the stochastic variational
+# Gaussian process (SVGP) model for efficient inference in large scale datasets.
+# For a basic introduction to the functionality of this library, please refer to
+# the [User Guide](@ref).
```
Comment (st--) [inline examples/regression/script.jl:8]: link doesn't seem to be working (same for classification example)

File: examples/regression/script.jl [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+using Plots
+default(; legend=:outertopright, size=(700, 400))
+
+using Random
+Random.seed!(1234)
+#md nothing #hide
+
+# ## Generate some training data
+#
+# The data generating function `g`
```
Comment (st--) [inline examples/regression/script.jl:26]:
```suggestion
# We define a data-generating function `g`:
```

File: examples/regression/script.jl [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+function g(x)
+    return sin(3π * x) + 0.3 * cos(9π * x) + 0.5 * sin(7π * x)
+end
```
Comment (st--) [inline examples/regression/script.jl:30]:
```suggestion
g(x) = sin(3π * x) + 0.3 * cos(9π * x) + 0.5 * sin(7π * x)
```
?

File: examples/regression/script.jl [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+# ## Set up a Flux model
+#
+# We shall use the excellent framework provided by [Flux.jl](https://fluxml.ai/)
+# to perform stochastic optimisation. The SVGP approximation has three sets of
+# parameters to optimise - the inducing input locations, the mean and covariance
+# of the variational distribution `q` and the parameters of the
+# kernel.
+#
+# First, we define a helper function to construct the kernel from its parameters
+# (often called kernel hyperparameters), and pick some initial values `k_init`.
```
Comment (st--) [inline examples/regression/script.jl:47]: It'd be very helpful to newcomers to add a comment to make explicit that Flux will do unconstrained optimisation, i.e. parameters anywhere on the real line, so we use `softplus()` to ensure positivity of kernel variances and lengthscales.

File: examples/regression/script.jl [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+
+using Flux
+
+function make_kernel(k)
+    return softplus(k[1]) * (SqExponentialKernel() ∘ ScaleTransform(softplus(k[2])))
```
Comment (st--) [inline examples/regression/script.jl:52]:
```suggestion
    return softplus(k[1]) * with_lengthscale(SqExponentialKernel(), softplus(k[2]))
```
?
or add a comment saying that `softplus(k[2])` is the _inverse_ lengthscale.

File: examples/regression/script.jl [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+end
```
Comment (st--) [inline examples/regression/script.jl:53]: Or how about:
```suggestion
function make_kernel(k_params)
    variance = softplus(k_params[1])
    lengthscale = softplus(k_params[2])
    return variance * with_lengthscale(SqExponentialKernel(), lengthscale)
end
```
(my preference)
or, if you prefer:

```suggestion
function make_kernel(k_params)
    variance = softplus(k_params[1])
    inv_lengthscale = softplus(k_params[2])
    return variance * (SqExponentialKernel() ∘ ScaleTransform(inv_lengthscale))
end
```

File: examples/regression/script.jl [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+
+k_init = [0.3, 10]
```
Comment (st--) [inline examples/regression/script.jl:55]: ideally, in a notebook you'd show how to set them to "interpretable" values - I don't know if there's an `inv_softplus` already implemented somewhere, but something along the lines of
```suggestion
init_variance = 1.3
init_lengthscale = 0.3
k_init = [inv_softplus(init_variance), inv_softplus(init_lengthscale)]
```

Comment (st--) [reply examples/regression/script.jl:55]: this would be much neater with bijectors ;)

File: examples/regression/script.jl [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+#md nothing #hide
+
+# Given a set of parameters, we now define a Flux 'layer' which forms the basis
+# of our model.
+
+struct SVGPModel
+    k  # kernel parameters
+    z  # inducing points
+    m  # variational mean
+    A  # variational covariance
```
Comment (st--) [inline examples/regression/script.jl:81]:
```suggestion
    A  # square-root of variational covariance
```

File: examples/regression/script.jl [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+# Then, we select some inducing input locations `z_init`. In this case, we simply choose
+# the first `M` data inputs.
+
+M = 50 # number of inducing points
+z_init = x[1:M]
+#md nothing #hide
+
+# Finally, we initialise the parameters of the variational distribution `q(u)`
+# where `u ~ f(z)` are the pseudo-points. We parameterise the covariance matrix
+# of `q` as `C = AᵀA` since this guarantees that `C` is positive definite.
```
Comment (st--) [inline examples/regression/script.jl:67]: Would inline-math work for the equations?

File: examples/regression/script.jl [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    m  # variational mean
+    A  # variational covariance
+end
+
+Flux.@functor SVGPModel (k, z, m, A)
+#md nothing #hide
+
+# Create the 'model' from the parameters - i.e. return the FiniteGP at inputs x,
+# the FiniteGP at inducing inputs z and the variational posterior over inducing
+# points - q(u).
```
Comment (st--) [inline examples/regression/script.jl:89]: This paragraph seems to hang a bit in the middle of nowhere - how does it connect to the previous/next ones?

File: examples/regression/script.jl [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+lik_noise = 0.3
+jitter = 1e-5
```
Comment (st--) [inline examples/regression/script.jl:92]:
```suggestion
jitter = 1e-5
#md nothing #hide
```

File: examples/regression/script.jl [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+#md nothing #hide
+
+# Finally, we initialise the parameters of the variational distribution `q(u)`
+# where `u ~ f(z)` are the pseudo-points. We parameterise the covariance matrix
+# of `q` as `C = AᵀA` since this guarantees that `C` is positive definite.
+
+m_init = zeros(M)
+A_init = Matrix{Float64}(I, M, M)
+q_init = MvNormal(m_init, A_init'A_init)
+#md nothing #hide
```
Comment (st--) [inline examples/regression/script.jl:72]: I think this would make more sense to the reader if you move it next to the `make_approx` definition - then it's all in one place, explanation, initialisation, and at-runtime evaluation

File: examples/regression/script.jl [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+
+# Return the loss given data - for the SVGP, the loss used is the negative ELBO
+# (also known as the Variational Free Energy). `n_data` is required for
+# minibatching used below.
+
+function loss(m::SVGPModel, x, y; n_data=length(y))
+    f = prior(m)
+    fx = f(x, lik_noise)
+    svgp = make_approx(m, f)
+    return -elbo(svgp, fx, y; n_data)
```
Comment (st--) [inline examples/regression/script.jl:136]: minor point: how about calling this a bit more explicitly `num_data`?

File: examples/regression/script.jl [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+data_loader = Flux.Data.DataLoader((x, y); batchsize=b)
+
+# The loss (negative ELBO) before training
+
+println(loss(model, x, y))
+
+# Train the model
+
+Flux.train!(
+    (x, y) -> loss(model, x, y; n_data=N),
```
Comment (st--) [inline examples/regression/script.jl:167]: Could you make more explicit that, when minibatching, it's crucial to explicitly pass the full length of the dataset (so that the scaling of data-fit vs KL penalty terms is correct)? maybe a comment here and a mention in the text above?

File: examples/regression/script.jl [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+    return softplus(k[1]) * (SqExponentialKernel() ∘ ScaleTransform(softplus(k[2])))
+end
+
+k_init = [0.3, 10]
+#md nothing #hide
+
+# Then, we select some inducing input locations `z_init`. In this case, we simply choose
+# the first `M` data inputs.
+
+M = 50 # number of inducing points
```
Comment (st--) [inline examples/regression/script.jl:61]:
```suggestion
M = 10 # number of inducing points
```
or something like that?

Would make the final plot less cluttered, and show the strength of the pseudo-point approximation

File: examples/regression/script.jl [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+    markeralpha=0.1,
+    xlim=(-1, 1),
+    xlabel="x",
+    ylabel="y",
+    title="posterior (VI with sparse grid)",
+    label="Train Data",
+)
+plot!(-1:0.001:1, post; label="Posterior")
+plot!(-1:0.001:1, g; label="True Function")
+vline!(z_init; label="Pseudo-points")
```
Comment (st--) [inline examples/regression/script.jl:196]: Could you change it to only mark the location as little ticks `|` on the x-axis (if you can get Plots.jl to let you overplot "outside the axes")? it looks quite cluttered at the moment

File: examples/regression/script.jl [hunk 17]
```diff
@@ -0,0 +1,196 @@
...
+    ncycle(data_loader, 300), # Train for 300 epochs
+    opt,
+)
+#md nothing #hide
+
+# Negative ELBO after training
+
+println(loss(model, x, y))
+
+# Finally, we plot samples from the optimised approximate posterior to see the
```
Comment (st--) [inline examples/regression/script.jl:178]: samples? isn't it the confidence intervals?

File: examples/regression/script.jl [hunk 18]
```diff
@@ -0,0 +1,196 @@
...
+# One of the major advantages of the SVGP model is that it allows stochastic
+# estimation of the ELBO by using minibatching of the training data. This is
+# very straightforward to achieve with `Flux.jl`'s utilities:
+
+b = 100 # minibatch size
+data_loader = Flux.Data.DataLoader((x, y); batchsize=b)
+
+# The loss (negative ELBO) before training
+
+println(loss(model, x, y))
```
Comment (st--) [inline examples/regression/script.jl:162]: Why not just
```suggestion
loss(model, x, y)
```
?

File: examples/regression/script.jl [hunk 19]
```diff
@@ -0,0 +1,196 @@
...
+# Gaussian process (SVGP) model for efficient inference in large scale datasets.
+# For a basic introduction to the functionality of this library, please refer to
+# the [User Guide](@ref).
+#
+# ## Setup
+
+using SparseGPs
+using Distributions
+using LinearAlgebra
+using IterTools
```
Comment (st--) [inline examples/regression/script.jl:15]: what are these used for? maybe an explicit
```suggestion
using IterTools: ncycle
```
(or whichever it is) could be helpful here

File: examples/classification/script.jl [hunk 20]
```diff
@@ -0,0 +1,184 @@
+# # Stochastic Variational Classification
+#
+# This example demonstrates how to carry out non-conjugate Gaussian process
+# inference using the stochastic variational Gaussian process (SVGP) model. For
+# a basic introduction to the functionality of this library, please refer to the
+# [User Guide](@ref).
```
Comment (st--) [inline examples/classification/script.jl:6]: again, reference doesn't work

File: examples/classification/script.jl [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+# First, we define the underlying latent GP
+# ```math
+# f \sim \mathcal{GP}(0, k(\cdot, \cdot'))
+# ```
+# and sample a function `f`.
+
+k_true = [30.0, 0.5]
+kernel_true = k_true[1] * (SqExponentialKernel() ∘ ScaleTransform(k_true[2]))
+
+lgp = LatentGP(GP(kernel_true), BernoulliLikelihood(), 1e-12)
```
Comment (st--) [inline examples/classification/script.jl:39]: would be helpful to explain the 1e-12 here again
```suggestion
jitter = 1e-12  # for numeric stability
lgp = LatentGP(GP(kernel_true), BernoulliLikelihood(), jitter)
```
along those lines..:)

File: examples/classification/script.jl [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+# latent GP with a Bernoulli likelihood. This results in a generative model that
+# we can use to produce some training data.
+#
+# First, we define the underlying latent GP
+# ```math
+# f \sim \mathcal{GP}(0, k(\cdot, \cdot'))
+# ```
+# and sample a function `f`.
+
+k_true = [30.0, 0.5]
```
Comment (st--) [inline examples/classification/script.jl:36]: could you find a more interesting sample? either a different seed or perhaps somewhat shorter lengthscale...

File: examples/classification/script.jl [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+# distribution with mean `μ`. We're only interested in the outputs at a subset
+# of inputs `x`, so we first pick some random input locations and then find the
+# corresponding values for `y`.
+
+N = 30  # The number of training points
+mask = sample(1:length(x_true), N; replace=false, ordered=true)  # Subsample some input locations
+x, y = x_true[mask], y_true[mask]
+
+scatter(x, y; label="Sampled outputs")
+plot!(x_true, mean.(lgp.lik.(f_true)); seriescolor="red", label="True function")
```
Comment (st--) [inline examples/classification/script.jl:61]:
```suggestion
plot!(x_true, mean.(lgp.lik.(f_true)); seriescolor="red", label="True probability")
```
or
```suggestion
plot!(x_true, mean.(lgp.lik.(f_true)); seriescolor="red", label="True mean")
```
seems more appropriate ("function" makes me think of the latent `f`)

File: examples/classification/script.jl [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+
+svgp_opt, f_opt = build_SVGP(final_params)
+post_opt = posterior(svgp_opt)
+l_post_opt = LatentGP(post_opt, BernoulliLikelihood(), jitter)
+
+post_f_samples = rand(l_post_opt.f(x_true, 1e-6), 20)
+post_μ_samples = mean.(l_post_opt.lik.(post_f_samples))
+
+plt = plot(x_true, post_μ_samples; seriescolor="red", linealpha=0.2, label="")
+scatter!(plt, x, y; seriescolor="blue", label="Data points")
```
Comment (st--) [inline examples/classification/script.jl:180]:
```suggestion
plt = scatter(plt, x, y; seriescolor="blue", label="Data points")
plot!(x_true, post_μ_samples; seriescolor="red", linealpha=0.2, label="Posterior samples")
```

**Review by st-- — APPROVED** [review]
Great start at the documentation! I've left a whole bunch of comments, but I'm happy for you to merge once you're happy with how you addressed them:) I've not reviewed the files that look copied from the other repos. I've been a bit less thorough about the classification notebook. I really like how you demonstrate both Flux.jl and Optim+ParameterHandling :)

File: examples/classification/script.jl [hunk 25]
```diff
@@ -0,0 +1,184 @@
+# # Stochastic Variational Classification
```
Comment (st--) [inline examples/classification/script.jl:1]: This isn't actually _stochastic_ optimisation...:
```suggestion
# # 2. Classification: Sparse Variational Approximation for Non-Conjugate Likelihoods
```

File: examples/regression/script.jl [hunk 26]
```diff
@@ -0,0 +1,196 @@
+# A recreation of <https://gpflow.readthedocs.io/en/master/notebooks/advanced/gps_for_big_data.html>
+
+# # Stochastic Variational Regression
```
Comment (st--) [inline examples/regression/script.jl:3]:
```suggestion
# # 1. Regression: Sparse Variational Gaussian Process for Stochastic Optimisation with Flux.jl
```

File: examples/classification/script.jl (hunk 25 above)
Comment (st--) [reply examples/classification/script.jl:1]: (NB- I'd prefer the more explicit headings even if you take out the numbering again)

File: examples/classification/script.jl [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+    return v, unflatten_PDmatrix
+end
+#md nothing #hide
+
+# Initialise the parameters
+
+M = 15  # number of inducing points
+raw_initial_params = (
+    k=(var=positive(rand()), precision=positive(rand())),
+    z=bounded.(range(0.1; stop=5.9, length=M), 0.0, 6.0),  # constrain z to simplify optimisation
```
Comment (st--) [inline examples/classification/script.jl:114]: how about:
```suggestion
    z=bounded.(range(0.1, 5.9; length=M), 0.0, 6.0),  # constrain z to simplify optimisation
```
feels a bit more natural to me

File: examples/regression/script.jl (hunk 1 above)
Comment (rossviljoen) [reply examples/regression/script.jl:8]: Yeah - I can't figure out why...

File: examples/classification/script.jl (hunk 22 above)
Comment (rossviljoen) [reply examples/classification/script.jl:36]: The reason for the simple sample is that optimisation gets a bit dodgy when it's more interesting :)

File: examples/regression/script.jl (hunk 10 above)
Comment (rossviljoen) [reply examples/regression/script.jl:89]: Ah - I forgot to remove this when moving things around

File: examples/classification/script.jl (hunk 22 above)
Comment (st--) [reply examples/classification/script.jl:36]: fair enough!

Comment (st--) [reply examples/classification/script.jl:36]: though I think that might be an issue with the parametrisation; normally, there shouldn't be any issues fitting 1D toy datasets!

Comment (st--) [reply examples/classification/script.jl:38]: https://github.com/JuliaGaussianProcesses/SparseGPs.jl/issues/35

File: examples/classification/script.jl (hunk 20 above)
Comment (st--) [reply examples/classification/script.jl:8]: #36

File: examples/regression/script.jl (hunk 1 above)
Comment (st--) [reply examples/regression/script.jl:9]: #36

File: examples/regression/script.jl [hunk 28]
```diff
@@ -0,0 +1,204 @@
...
+
+function make_kernel(k_params)
+    variance = StatsFuns.softplus(k_params[1])
+    lengthscale = StatsFuns.softplus(k_params[2])
+    return variance * with_lengthscale(SqExponentialKernel(), lengthscale)
+end
+
+init_variance = 1.3
+init_lengthscale = 0.3
+k_init = [invsoftplus(init_variance), invsoftplus(init_lengthscale)]
```
Comment (st--) [inline examples/regression/script.jl:62]: oh, that IS a thing :D

File: src/elbo.jl [hunk 29]
```diff
@@ -7,33 +7,33 @@ ScalarLikelihood = Union{
     GammaLikelihood,
 }
 
-abstract type ExpectationMethod end
-struct Default <: ExpectationMethod end
-struct Analytic <: ExpectationMethod end
+abstract type QuadratureMethod end
+struct Default <: QuadratureMethod end
```
Comment (st--) [inline src/elbo.jl:11]:
```suggestion
struct DefaultQuadrature <: QuadratureMethod end
```
